### PR TITLE
Add unit tests

### PR DESCRIPTION
**Pull request by JeremyWesthead**
By no means the end of adding unit tests, but I'm aware this is already 5,000+ lines so thought it would be best to split into multiple PRs.
Part of these changes are also appeasing `clippy` (Rust's built in linter)

### CODE REVIEW

**Comment by mcolpus** [comment]
I'll start taking a look at this

**Comment by mcolpus** [comment]
You can add `#![allow(clippy::needless_return)]` to the top of lib.rs and you then don't need to use implicit returns everywhere

File: src/genome.rs [hunk 1]
```diff
@@ -467,12 +469,4073 @@ pub fn mutate(reference: &Genome, vcf: VCFFile) -> Genome {
...
...cagcagtggggaatattgcacaatgggcgcaagcctgatgcagcgacgccgcgtgggggatgacggccttcgggttgtaaacctctttcaccatcgacgaaggtccgggttctctcggattgacggtaggtggagaagaagcaccggccaactacgtgccagcagccgcggtaatacgtagggtgcgagcgttgtccggaattactgggcgtaaagagctcgtaggtggtttgtcgcgttgttcgtgaaatctcacggcttaactgtgagcgtgcgggcgatacgggcagactagagtactgcaggggagactggaattcctggtgtagcggtggaatgcgcagatatcaggaggaacaccggtggcgaaggcgggtctctgggcagtaactgacgctgaggagcgaaagcgtggggagcgaacaggattagataccctggtagtccacgccgtaaacggtgggtactaggtgtgggtttccttccttgggatccgtgccgtagctaacgcattaagtaccccgcctggggagtacggccgcaaggctaaaactcaaaggaattgacgggggcccgcacaagcggcggagcatgtggattaattcgatgcaacgcgaagaaccttacctgggtttgacatgcacaggacgcgtctagagataggcgttcccttgtggcctgtgtgcaggtggtgcatggctgtcgtcagctcgtgtcgtgagatgttgggttaagtcccgcaacgagcgcaacccttgtctcatgttgccagcacgtaatggtggggactcgtgagagactgccggggtcaactcggaggaaggtggggatgacgtcaagtcatcatgccccttatgtccagggcttcacacatgctacaatggccggtacaaagggctgcgatgccgcgaggttaagcgaatccttaaaagccggtctcagttcggatcggggtctgcaactcgaccccgtgaagtcggagtcgctagtaatcgcagatcagcaacgctgcggtgaatacgttcccgggccttgtacacaccgcccgtcacgtcatgaaagtcggtaacacccgaagccagtggcctaaccctcgggagggagctgtcgaaggtgggatcggcgattgggacgaagtcgtaacaaggtagccgtaccggaaggtgcggctggatcacctcctttct".to_string()
```
Comment (mcolpus) [inline src/genome.rs:529]: Are the two collects needed?

File: src/genome.rs [hunk 2]
```diff
@@ -467,12 +469,4073 @@ pub fn mutate(reference: &Genome, vcf: VCFFile) -> Genome {
...
+        );
+    }
+
+    #[test]
+    fn test_invalid_args() {
+        assert_panics!(Genome::new("not/a/path"));
```
Comment (mcolpus) [inline src/genome.rs:554]: do you want to have some catch to check the panic message is correct?

File: src/genome.rs [hunk 3]
```diff
@@ -467,12 +469,4073 @@ pub fn mutate(reference: &Genome, vcf: VCFFile) -> Genome {
...
+            } else if idx - offset < 30 {
+                assert_eq!(pos.reference, 'g');
+            } else if idx - offset < 40 {
+                assert_eq!(pos.reference, 't');
+            } else {
+                assert_eq!(pos.reference, 'a');
+                offset = idx;
+            }
+
+            if (0..30).filter(|x| x == &idx).count() > 0 {
```
Comment (mcolpus) [inline src/genome.rs:578]: Why the strange use of iterator? Is this doing the same as `if 0 <= idx <= 30`?

Comment (mcolpus) [reply src/genome.rs:578]: and should it be 4..30 to match the file?

File: src/genome.rs [hunk 4]
```diff
@@ -467,12 +469,4073 @@ pub fn mutate(reference: &Genome, vcf: VCFFile) -> Genome {
...
+            if (91..99).filter(|x| x == &idx).count() > 0 {
+                assert!(pos.genes.contains(&"C".to_string()));
+            }
+        }
+    }
+
+    #[test]
+    fn test_dna_genes() {
+        let mut genome = Genome::new("reference/TEST-DNA.gbk");
+        let gene_a = genome.get_gene("A".to_string());
```
Comment (mcolpus) [inline src/genome.rs:593]: I'm not sure why the `get_gene` function doesn't just take a &str? As far as I can see it either uses a borrow or clones, so don't need to take ownership

File: src/genome.rs [hunk 5]
```diff
@@ -467,12 +469,4073 @@ pub fn mutate(reference: &Genome, vcf: VCFFile) -> Genome {
...
+            },
+        ];
+
+        for (idx, variant) in genome_diff.variants.iter().enumerate() {
+            assert_eq!(*variant, expected_genome_variants[idx]);
+        }
+
+        let expected_genome_minor_variants = vec![
+            Variant {
+                variant: "4730t>t:99".to_string(),
```
Comment (mcolpus) [inline src/genome.rs:787]: why is it t>t when the ref is a, and the alt is t?

File: src/genome.rs [hunk 6]
```diff
@@ -467,12 +469,4073 @@ pub fn mutate(reference: &Genome, vcf: VCFFile) -> Genome {
...
+                },
+                vcf_idx: Some(1),
+                indel_length: 2,
+                indel_nucleotides: Some("ca".to_string()),
+                gene_position: None,
+                codon_idx: None,
+                gene_name: None,
+            },
+            Variant {
+                variant: "69g>x".to_string(),
```
Comment (mcolpus) [inline src/genome.rs:1740]: Why does this result in an x? Is it because alt==ref?

File: src/vcf.rs [hunk 7]
```diff
@@ -515,28 +550,52 @@ impl VCFFile {
...
 ///
 /// # Returns
 /// SNP distance between the two strings
-fn snp_dist(reference: &String, alternate: &String) -> i64 {
+fn snp_dist(reference: &str, alternate: &str) -> i64 {
+    if reference.len() != alternate.len() {
+        panic!("Reference and alternate strings are not the same length!");
+    }
     let mut dist = 0;
     for i in 0..reference.len() {
```
Comment (mcolpus) [inline src/vcf.rs:600]: I'm not sure you'll want to take the length of a str like that (https://stackoverflow.com/questions/46290655/does-rusts-string-have-a-method-that-returns-the-number-of-characters-rather-th)
I imagine it's fine here as we know they are normal utf8 characters.
I would consider rewriting this with a zip of iterators:
```
fn count_differences(s1: &str, s2: &str) -> usize {
    s1.chars().zip(s2.chars()).filter(|(c1, c2)| c1 != c2 && c1 != 'N' && c2 != 'N').count()
}
```

File: src/vcf.rs [hunk 8]
```diff
@@ -545,5 +604,730 @@ fn snp_dist(reference: &String, alternate: &String) -> i64 {
...
+
+        assert_eq!(
+            VCFFile::simplify_call("CCC".to_string(), "AAAAA".to_string()),
+            vec![
+                (2, AltType::INS, "AA".to_string()),
+                (0, AltType::SNP, "A".to_string()),
+                (1, AltType::SNP, "A".to_string()),
+                (2, AltType::SNP, "A".to_string()),
+            ]
+        );
```
Comment (mcolpus) [inline src/vcf.rs:708]: I'd probably want to add a test case like ACGT -> ACGAGT or something where the ins/del is in the middle

File: src/vcf.rs [hunk 9]
```diff
@@ -545,5 +604,730 @@ fn snp_dist(reference: &String, alternate: &String) -> i64 {
...
+                    ("COV".to_string(), vec!["1".to_string(), "5".to_string()]),
+                ]),
+                is_filter_pass: true,
+            };
+            let (calls, minor_calls) = VCFFile::parse_record_for_calls(record, 3);
+            assert_eq!(calls.len(), 1);
+            assert_eq!(minor_calls.len(), 0);
+            assert_eq!(calls[0].genome_index, 2);
+            assert_eq!(calls[0].reference, "t".to_string());
+            assert_eq!(calls[0].alt, "t".to_string());
```
Comment (mcolpus) [inline src/vcf.rs:929]: It's a little funny that "t" is the alt? Is this a particularity to dels?

**Review by mcolpus — APPROVED** [review]
Generally looks good to me. Tried to go through it fairly closely, but there were a lot of tests. All the reformatting for clippy looks good. I've not properly checked the logic of the src code in much detail (as it wasn't part of this PR), but I'm relying on the tests for picking up errors with that.
This must of taken ages to write!

File: src/genome.rs (hunk 5 above)
Comment (mcolpus) [reply src/genome.rs:787]: Is this to do with the vcf not matching the ref file properly?

File: src/main.rs [hunk 10]
```diff
@@ -73,9 +72,9 @@ fn main() {
 
     let gene_start = SystemTime::now();
     for gene_name in sample.genes_with_mutations.clone().iter() {
-        if gene_name != target_gene {
-            continue;
-        }
+        // if gene_name != target_gene {
+        //     continue;
```
Comment (mcolpus) [inline src/main.rs:76]: Do all these commented sections want to be left in??

File: src/genome.rs (hunk 1 above)
Comment (JeremyWesthead) [reply src/genome.rs:529]: One's for collecting the nc sequence to a `Vec<char>` then the second is for collecting the return into a `String` for comparison. Only did the latter as I had a string of the nucleotide sequence; I've avoided this for tests I added today

File: src/genome.rs (hunk 2 above)
Comment (JeremyWesthead) [reply src/genome.rs:554]: Not massively concerned; I'm more checking for lack of undefined behaviour than error messages as that should be caught and handled in Python

File: src/genome.rs (hunk 3 above)
Comment (JeremyWesthead) [reply src/genome.rs:578]: That's something I got 'suggested' by clippy. It's doing the same thing

And nope, once promoter regions are assigned the `A` gene expands back to genome position 0

File: src/genome.rs (hunk 5 above)
Comment (JeremyWesthead) [reply src/genome.rs:787]: Good spot! Not sure... Writing so many of these means some things slip through

Comment (JeremyWesthead) [reply src/genome.rs:787]: Ohh nevermind. The VCF is dummy data and doesn't actually match the genome at that position. VCF part is reporting correctly that the VCF claims a reference of `c`, and the variant is correctly reporting that the genbank file has a reference of `t`

File: src/genome.rs (hunk 6 above)
Comment (JeremyWesthead) [reply src/genome.rs:1740]: It's technically a null call according to the coverage on the GT.
`0/0`, but only 1 read supporting, so is nulled

File: src/vcf.rs (hunk 7 above)
Comment (JeremyWesthead) [reply src/vcf.rs:600]: Yeah I saw that, but honestly if a VCF has non UTF-8 characters in it there's going to be larger issues than just that so I think simplistic code is better here

File: src/vcf.rs (hunk 8 above)
Comment (JeremyWesthead) [reply src/vcf.rs:708]: It's less clear as it's repeats of the same base, but indels mid-way through are included in tests below https://github.com/oxfordmmm/grumpy/blob/4dfe87ea033ef6fdf3dd19d541144fa0f63abe0b/src/vcf.rs#L690

File: src/vcf.rs (hunk 9 above)
Comment (JeremyWesthead) [reply src/vcf.rs:929]: With indels, the alt is actually used to store the inserted/deleted bases. The different interpretations are checked based on the AltType
I agree though, single base indels can look confusing with this paradigm, but this is documented with a docstring here https://github.com/oxfordmmm/grumpy/blob/4dfe87ea033ef6fdf3dd19d541144fa0f63abe0b/src/common.rs#L82 which makes it to the crate docs neatly https://docs.rs/grumpy/0.1.10/grumpy/common/struct.Evidence.html#structfield.alt

File: src/main.rs (hunk 10 above)
Comment (JeremyWesthead) [reply src/main.rs:76]: I use them for testing as it makes it easy to pull out gene-level variation (but usually want the full genome). Commented as they're in frequent but intermittent use and I'll clean them up once I think all edge cases have been covered

File: src/genome.rs (hunk 4 above)
Comment (JeremyWesthead) [reply src/genome.rs:593]: It probably should, I'll fix

File: src/vcf.rs (hunk 7 above)
Comment (mcolpus) [reply src/vcf.rs:600]: Very true. I think I just like the zip as it avoids all the unwrapping.

File: src/vcf.rs (hunk 8 above)
Comment (mcolpus) [reply src/vcf.rs:708]: though the indel is actually put at the end. So it's 3 snps and then an insertion

File: src/genome.rs (hunk 4 above)
Comment (JeremyWesthead) [reply src/genome.rs:593]: I lied. It's actually not quite that simple; `get_gene` inserts into a HashMap indexed on String so it is actually owned by the object (hence the clones). Changing the HashMap to use `&str` causes even more issues - can't use `&str` as python classes don't support lifetime parameters; can't use `str` as it doesn't implement `Sized` etc.
Basically, it's a lot more effort to change given the Python bindings are agnostic to this anyway

File: src/vcf.rs (hunk 8 above)
Comment (JeremyWesthead) [reply src/vcf.rs:708]: Ah true, added on my other branch as I forgot I wasn't on this one

**Pull request merged**